### DISCUSSION TITLE
On disconnect (close SQL editor tab) I am having functions removed from the DB.

### DISCUSSION THREAD

**Discussion by rjmurray54** [dbeaver/dbeaver#11492] [Q&A] [answered]
#### System information: 
- Windows/10 64-bit -  Version: 1909 Build: 18363.1256
- DBeaver version: 7.3.0.202011291229

#### Connection specification:
- PostgreSQL 13.1 

#### Describe your question:
New to DBeaver.  If I open a new SQL Editor tab and build a function and can execute it and see it in the function list.  But when I close the SQL Editor Tab, the function is deleted from the Database.  I know there must be a setting somewhere to keep DBeaver from removing the function, but I will be damn if I can find it . . . HELP PLEASE!!  Thanks in advance!

**Comment by kseniiaguzeeva** [comment]
Can you be in manual commit mode? Please check.
If I am right, change it on auto-commit or press commit after query execution to apply it.
![dbeaver_Vu3S2tKKio](https://user-images.githubusercontent.com/51405061/102771795-90006600-4397-11eb-9cf9-74d3e05d7830.png)

**Comment by rjmurray54** [comment]
Hi Kseniia,

I did exactly like you said,

  1.  I set it in Manual Commit mode,
  2.  compiled the function,
  3.  committed the outstanding,
  4.  then refreshed the functions in the treeview, in the database I am working in
  5.  the function was listed in the Functions,
  6.  then I closed the tab and refreshed again
  7.  and the function was gone!!


Regards,

**Comment by rjmurray54** [comment]
And as a follow-up, no I was in AUTO commit mode . . .!

Thanks,

**Comment by rjmurray54** [comment]
I clicked send to quickly, if you would like to see it happen, we can do a ZOOM session and I can share my screen . . . just let me know.

Thanks,

**Comment by kseniiaguzeeva** [comment]
Let's try to look at video at first. I can create the function and it is saved. Please send me video how do you do it?
https://user-images.githubusercontent.com/51405061/102862213-cb576f00-4441-11eb-9fe4-1a28f00ef4b4.mov

**Comment by rjmurray54** [comment]
Hi Kseniia,

Please look at the readme and there are two videos, look at part 1 and then part 2

https://github.com/rjmurray54/Disappearing-Function-Issue-10713

**Comment by kseniiaguzeeva** [comment]
Still cannot reproduce. Please try to create the view and open it.
```
CREATE OR REPLACE VIEW show_functions AS
    SELECT routine_name FROM information_schema.routines 
        WHERE routine_type='FUNCTION' AND specific_schema='tryon';
```
Is the function displayed there?

**Comment by rjmurray54** [comment]
Hi Kseniia,

Nope, it is not there, in the view!

**Comment by kseniiaguzeeva** [comment]
Please send me your DDL example and log files https://github.com/dbeaver/dbeaver/wiki/Log-files

**Comment by rjmurray54** [comment]
Good Day Kseniia,

Attached, are what I hope is everything you need to reproduce the issue, if not just let me know and I will get you what you need.  Thanks for all of the help, it is greatly appreciated.

Happy Holidays and stay safe!

**Comment by kseniiaguzeeva** [comment]
I do not see any attachments.. Please send them one more time

**Comment by rjmurray54** [comment]
There should be a Case_#_10713.RAR attached . . .

**Comment by rjmurray54** [comment]
Hi Kseniia,

I sent the attachment in the prior email, it was only 28Kb so it should not get stripped from them email for size.  Is does have an SQL extension, I don’t know if you email will dump that off or not . . . let me know and we will figure something else out.

**Comment by p91paul** [comment]
You can comment this thread by replying to the github email, but github is still not a mailing list. So it's likely it doesn't support email attachments. You are probably better off posting your comment and attaching stuff through the web page instead of using email.

**Comment by rjmurray54** [comment]
[Case_#_10713.Docx](https://github.com/dbeaver/dbeaver/files/5747193/Case_._10713.Docx)
I had to change the extension to DOCX, but it is really a RAR.  So everything is in that compresses file.  Thanks!  Rick

**Comment by rjmurray54** [comment]
Please check the entry I put on GITHub . . .

**Comment by kseniiaguzeeva** [comment]
The files are extracted.  I will see them and answer

**Comment by kseniiaguzeeva** [comment]
Please send me also DDL for types from the attached example. I cannot create several tables because some custom types do not exist.

**Comment by rjmurray54** [comment]
Good Morning . . . There was no attachment!

**Comment by kseniiaguzeeva** [comment]
I mean this one https://github.com/dbeaver/dbeaver/issues/10713#issuecomment-751687394

**Comment by rjmurray54** [comment]
[Tryon_DB_Dumpfile.zip](https://github.com/dbeaver/dbeaver/files/5750878/Tryon_DB_Dumpfile.zip)

It was easier to just send you a dump of the DB!!  Thanks for all of your Help!!

**Comment by rjmurray54** [comment]
The password for the DB is "jimmydizzle1"

**Comment by rjmurray54** [comment]
Hi Kseniia,

Happy New Year my friend!  I was just checking to see if you had made any progress with my issue?

**Comment by rjmurray54** [comment]
Hi Kseniia,

I sent you an email about a week ago to see what the status of this issue was and I did not hear back from you.  I sent you that dump of our database back during the holidays and since then I have not heard anything from you.  It is getting to the point where this will become a major issue, so please get back as soon as you can.

**Comment by kseniiaguzeeva** [comment]
Thank you for your reminding. Yesterday I discussed with the colleague your issue. And we still do not have an idea how to reproduce it.

Also I cannot restore your database, because some types are missed from DDL of the dump file too. Let's go from another side. Try to create a simple function (or another from [PostgreSQL documentation](https://www.postgresql.org/docs/9.5/sql-createfunction.html)):
```
CREATE FUNCTION add(integer, integer) RETURNS integer
    AS 'select $1 + $2;'
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT;
```

Will it be saved or disappear?

**Comment by rjmurray54** [comment]
Good Day Kseniia,

Thanks so much for getting back, because this is starting to become a major issue and I just don’t know where to turn. . .

I did create the below function and it does NOT go away, when closing DBeaver or when closing an SQL Editor Tab.  The other thing that I failed to tell you, is the local copy of the DB that I am using, is running in a Docker container . . . which should not have any impact, but I wanted you to know.

**Comment by kseniiaguzeeva** [comment]
I really want to find a reason, but I cannot do it without your help.
- Can you reproduce it in 7.3.2?
- Please give me a DDL example which I can run and create your environment.
- Is it the only function which disappears? Does it always disappear?
- Could you tell the exact steps to reproduce one more time (please do not forget any details like about docker etc.)
- Can you reproduce the issue on another database?

**Comment by rjmurray54** [comment]
I am running Version 7.3.2.202101032114 of DBeaver


  *   Can you reproduce it in 7.3.2?

I am running Version 7.3.2.202101032114 of DBeaver and yes it still happens in this version.

  *   Please give me a DDL example which I can run and create your environment.
Below is a function that does not need database tables, you do not need to run it.  Simply perform the following steps:
                Start a docker container that has any database in it
                Open the above version of DBeaver and navigate to the database running in Docker
                Open a SQL Editor tab
                Paste the below function code into the Editor tab
                Create the temp table from the code in the function
                Create the function
                Refresh the functions branch in the treeview of the database you are working with.
                Close your open copy of DBeaver
                Reopen DBeaver
                Go to the Functions branch and see if the function is still there . . . mine is NOT!!

  *   Is it the only function which disappears? Does it always disappear?
No it is not the only function that goes away when I close DBeaver, but both functions use TEMP tables, I don’t know if it has anything to do with it.
CREATE OR REPLACE FUNCTION tryon.uf_rjm_test_Func(
in
       in_appt_sts         varchar,     -- values can be 'A'= All, 'B'= Booked, or 'O'= Open
       in_tm_frm           varchar,     -- what timeframes to look for
       in_loc_id           bigint,
       in_provider bigint,
       in_appt_str_ts      timestamp,
       in_appt_end_ts      timestamp,
       in_duration int
)

/* -------------------------------------------------------------------
* Start the first sequence of loading all times with "Open" times
* the second sequence will update the rows with any appointments
* that are currently scheduled for the requested location and provider
* then the temp table will be returned to the caller.
 *-------------------------------------------------------------------- */

RETURNS SETOF tb_tmpAppntmnts as

$Function$
DECLARE
    tb_rjm_test record;

/*DECLARE
    OpnAppt record;   */

DECLARE
    wk_str_ts              timestamp;
    wk_end_ts              timestamp;
    wk_stop_ts             timestamp;

Begin
       CREATE temp TABLE IF NOT EXISTS tb_tmpAppntmnts(
              tmp_id              serial       not null,
              patient_id          bigint,
              reason_id           bigint,
              staff_id            bigint,
              intake_notes text,
              is_online           boolean,
              appt_str_ts         timestamp,
              appt_end_ts         timestamp,
              loc_id                     bigint,
              pat_name            text,
              apptmt_reason varchar,
              loc_name            varchar,
              loc_addr        text,
              CONSTRAINT tmpAppntmnts_primary_key PRIMARY KEY (tmp_id)
              );

              truncate table tb_tmpAppntmnts;

             wk_str_ts    := in_appt_str_ts;
             wk_end_ts    := in_appt_end_ts;
              wk_stop_ts   := in_appt_end_ts;
              wk_end_ts    := ( wk_str_ts::timestamp + in_duration * interval '1 minute' ) ;

/*  Check the input parm for what to return to caller
*  --------------------------------------------------
*/

Case in_appt_sts

       When 'O' Then -- Fetch and display OPEN appointment times

       /* Loop through creating open time slots for this Location and Provider
        *--------------------------------------------------------------------- */
           Loop

                  insert into tb_tmpAppntmnts( patient_id, reason_id, staff_id,           intake_notes,       is_online,   appt_str_ts, appt_end_ts, loc_id,      pat_name,        apptmt_reason, loc_name, loc_addr )
                                                        values( null,       null,      in_provider, null,               null,               wk_str_ts,   wk_end_ts,   in_loc_id, 'Open',      '', '', ''  );

                     wk_str_ts := wk_end_ts ;
                     wk_end_ts := (wk_end_ts::timestamp + in_duration * interval '1 minute' );

                     if wk_end_ts > wk_stop_ts then
                                  Exit;
                     end if;

              End Loop ;


End case;

/* Return the below results, only If it is an "O" option,
* used to remove booked appointment times.
* -----------------------------------------------------
*/

Return Query
       Select Appt.tmp_id, Appt.patient_id, Appt.reason_id, Appt.staff_id, Appt.intake_notes, Appt.is_online, Appt.appt_str_ts, Appt.appt_end_ts, Appt.loc_id, Appt.pat_name, Appt.apptmt_reason, loc.name, (addr.street1 || ' ' || addr.street2 || ' ' || addr.city || addr.state || ' ' || addr.zip) as loc_addr
       From  tb_tmpAppntmnts Appt
        Where appt.patient_id is NULL
       order by appt_str_ts asc;

       RETURN;
END;

$Function$
LANGUAGE plpgsql;


End of Function ----------------------------------------------------------------------------------------------------------------------

**Comment by rjmurray54** [comment]
Hi Kseniia,

Well we figured out what is causing it.  It seems that when it does the removal of the temp table in that code, that it is doing a cascade delete, so it removes the function as well!!  What we did was make the temp table a regular table entity and not temp and when we close DBeaver, the function is the when we return.  So, that should give you guys something to look at!

Please get back with any comments or possibilities of getting this fixed . . .

**Comment by kseniiaguzeeva** [comment]
Steps to reproduce:
1. Insert in SQL Editor (e.g. for PostgreSQL 12 dvdrental.tryon) the script from the comment above
2. execute only it from the script at first:

```
CREATE temp TABLE IF NOT EXISTS tb_tmpAppntmnts(
              tmp_id              serial       not null,
              patient_id          bigint,
              reason_id           bigint,
              staff_id            bigint,
              intake_notes text,
              is_online           boolean,
              appt_str_ts         timestamp,
              appt_end_ts         timestamp,
              loc_id                     bigint,
              pat_name            text,
              apptmt_reason varchar,
              loc_name            varchar,
              loc_addr        text,
              CONSTRAINT tmpAppntmnts_primary_key PRIMARY KEY (tmp_id)
              );
```
3. Refrech and check Functions folder> the function is in it
4. Relaunch DBeaver and open the folder Functions again

Expected: the function is displayed there
Actual result: the function disappeared

**Comment by rjmurray54** [comment]
That is not the issue at all, if you create a function with a temp table in the script, when run it and then close DBeaver and reopen Dbeaver, the function has been removed.  We have nailed the problem down to the fact that when the Temp Table is removed, it is doing a CASCADE delete, which also removes the function and it should NOT, it should just remove any temp tables that are created during the session.

**Comment by ShadelessFox** [comment, marked as answer]
Hello @rjmurray54,

This is not a DBeaver problem - this is how PostgreSQL works. **Temporary tables are dropped at the end of the session, along with all objects that depend on it**. The solution is (that, in fact, was already mentioned) is to make the table non-temporary or create custom type instead of table.

https://www.postgresql.org/docs/9.3/sql-createtable.html

**Comment by rjmurray54** [comment]
Dear Sir / Madam,

Thank you for responding to my perceived issue with DBeaver.  I come from a SQL Server background and the SQL Server -vs- PostgreSQL handles temporary tables or rather objects that are associated with temp tables, totally different.  When a temp table, whether it be local or global, is deleted at the end of a session, only the table is removed and not ALL objects associated to it.  Hence my confusion in thinking it was an error with DBeaver, my bad.

I do have one final question, is regard to your response.  What I did not understand after reading your response, was how creating a custom type would help resolve this functionality, in PostgreSQL.  I read the documentation you referenced in your response and did not see how that might help.  If you would be so kind as to explain, in a little more detail how this might help, it would be greatly appreciated.

Thank you in advance for taking a little time to help me understand your comment.  And then believe it or not, I really will quit pestering you about this topic.

**Comment by ShadelessFox** [comment]
I'm sorry for confusing you by mentioning the ability to use the custom type in this case. I haven't noticed that there's a loop in your function that inserts values into the table, hence may result in multiple rows. I was thinking about using custom type as a _tuple_ to store several values in instead of utilizing the table, but that's not your case, because it would be only suitable for one _row_. I hope this explanation makes sense.

**Comment by rjmurray54** [comment]
That helped a lot thanks!!
